File: src/main/go/leetcode/editor/cn/letter_combinations_of_a_phone_number/LetterCombinationsOfAPhoneNumber.go
//17.letter-combinations-of-a-phone-number
//给定一个仅包含数字 2-9 的字符串，返回所有它能表示的字母组合。答案可以按 任意顺序 返回。
//
// 给出数字到字母的映射如下（与电话按键相同）。注意 1 不对应任何字母。
//
//
//
//
//
// 示例 1：
//
//
//输入：digits = "23"
//输出：["ad","ae","af","bd","be","bf","cd","ce","cf"]
//
//
// 示例 2：
//
//
//输入：digits = ""
//输出：[]
//
//
// 示例 3：
//
//
//输入：digits = "2"
//输出：["a","b","c"]
//
//
//
//
// 提示：
//
//
// 0 <= digits.length <= 4
// digits[i] 是范围 ['2', '9'] 的一个数字。
//
// Related Topics 深度优先搜索 递归 字符串 回溯算法
// 👍 1341 👎 0

package main

import (
	"fmt"
	"strings"
)

//leetcode submit region begin(Prohibit modification and deletion)

var keys = []string{
	"abc",
	"def",
	"ghi",
	"jkl",
	"mno",
	"pqrs",
	"tuv",
	"wxyz",
}

func letterCombinations(digits string) []string {
	var ans []string
	var backtrack func(idx int, digits string, res []string)
	backtrack = func(idx int, digits string, res []string) {
		if idx == len(digits) {
			ans = append(ans, strings.Join(res, ""))
			return
		}
		for _, s := range keys[digits[idx]-'0'-2] {
			res[idx] = string(s)
			backtrack(idx+1, digits, res)
		}
	}
	backtrack(0, digits, make([]string, len(digits)))
	return ans
}

//leetcode submit region end(Prohibit modification and deletion)

// test from here
func main() {
	fmt.Println(letterCombinations("23"))
}

File: src/main/java/leetcode/editor/en/LetterCombinationsOfAPhoneNumber.java
//17.letter-combinations-of-a-phone-number
/**
 * Given a string containing digits from 2-9 inclusive, return all possible letter
 * combinations that the number could represent. Return the answer in any order.
 * <p>
 * A mapping of digits to letters (just like on the telephone buttons) is given
 * below. Note that 1 does not map to any letters.
 * <p>
 * <p>
 * Example 1:
 * <p>
 * <p>
 * Input: digits = "23"
 * Output: ["ad","ae","af","bd","be","bf","cd","ce","cf"]
 * <p>
 * <p>
 * Example 2:
 * <p>
 * <p>
 * Input: digits = "2"
 * Output: ["a","b","c"]
 * <p>
 * <p>
 * <p>
 * Constraints:
 * <p>
 * <p>
 * 1 <= digits.length <= 4
 * digits[i] is a digit in the range ['2', '9'].
 * <p>
 * <p>
 * Related Topics Hash Table String Backtracking 👍 20910 👎 1131
 */

package leetcode.editor.en;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class LetterCombinationsOfAPhoneNumber {
    public static void main(String[] args) {
        Solution solution = new LetterCombinationsOfAPhoneNumber().new Solution();
    }

    //leetcode submit region begin(Prohibit modification and deletion)
    class Solution {
        private static final Map<Character, char[]> MAP = Map.ofEntries(
                Map.entry('2', new char[]{'a', 'b', 'c'}),
                Map.entry('3', new char[]{'d', 'e', 'f'}),
                Map.entry('4', new char[]{'g', 'h', 'i'}),
                Map.entry('5', new char[]{'j', 'k', 'l'}),
                Map.entry('6', new char[]{'m', 'n', 'o'}),
                Map.entry('7', new char[]{'p', 'q', 'r', 's'}),
                Map.entry('8', new char[]{'t', 'u', 'v'}),
                Map.entry('9', new char[]{'w', 'x', 'y', 'z'})
        );

        public List<String> letterCombinations(String digits) {
            List<String> list = new ArrayList<>();
            if(digits!=null&&digits.length()>0){
                dfs(digits, 0, new char[digits.length()], list);
            }
            return list;
        }

        private void dfs(String digits, int index, char[] path, List<String> res) {
            if (index == digits.length()) {
                res.add(new String(path));
                return;
            }
            char ch = digits.charAt(index);
            for (char c : MAP.getOrDefault(ch, new char[0])) {
                path[index] = c;
                dfs(digits, index + 1, path, res);
            }
        }
    }
//leetcode submit region end(Prohibit modification and deletion)

}